client-flink/flink-1.14/src/main/java/org/apache/celeborn/plugin/flink/ShuffleResourceTracker.java

```diff
@@ -0,0 +1,167 @@
+  private static final int MAX_RETRY_TIMES = 3;
+
+  public ShuffleResourceTracker(
+      ExecutorService executorService, LifecycleManager lifecycleManager) {
+    this.executorService = executorService;
+    ShuffleWorkerStatusListener shuffleWorkerStatusListener =
+        new ShuffleWorkerStatusListener(lifecycleManager, shuffleResourceListeners);
+    lifecycleManager.registerWorkerStatusListener(shuffleWorkerStatusListener);
+  }
+
+  public void registerJobContext(JobShuffleContext jobShuffleContext) {
+    jobShuffleContextMap.put(jobShuffleContext.getJobId(), jobShuffleContext);
+  }
+
+  public void addPartitionResource(
+      JobID jobId, int shuffleId, int partitionId, ResultPartitionID partitionID) {
+    JobShuffleContext jobShuffleContext = jobShuffleContextMap.get(jobId);
+    ShuffleResourceListener shuffleResourceListener =
+        shuffleResourceListeners.computeIfAbsent(
+            shuffleId,
+            (s) -> new ShuffleResourceListener(jobShuffleContext, shuffleId, executorService));
+    shuffleResourceListener.addPartitionResource(partitionId, partitionID);
+  }
+
+  public void removePartitionResource(int shuffleId, int partitionId) {
+    ShuffleResourceListener shuffleResourceListener = shuffleResourceListeners.get(shuffleId);
+    if (shuffleResourceListener != null) {
+      shuffleResourceListener.removePartitionResource(partitionId);
+    }
+  }
+
+  public void removeShuffleListener(int shuffleId) {
+    shuffleResourceListeners.remove(shuffleId);
+  }
+
+  public void unRegisterJob(JobID jobID) {
+    jobShuffleContextMap.remove(jobID);
+  }
+
+  public static class ShuffleResourceListener {
+
+    private final JobShuffleContext context;
+    private final int shuffleId;
+    private final ExecutorService executorService;
+
+    // celeborn partitionId -> Flink ResultPartitionID
+    private Map<Integer, ResultPartitionID> resultPartitionMap = new ConcurrentHashMap<>();
+
+    public ShuffleResourceListener(
+        JobShuffleContext jobShuffleContext, int shuffleId, ExecutorService executorService) {
+      this.context = jobShuffleContext;
+      this.shuffleId = shuffleId;
+      this.executorService = executorService;
+    }
+
+    public void addPartitionResource(int partitionId, ResultPartitionID partitionID) {
+      resultPartitionMap.put(partitionId, partitionID);
+    }
+
+    public void notifyStopTrackingPartitions(Set<Integer> partitionIds) {
+      Set<ResultPartitionID> resultPartitionIDS = new HashSet<>();
+
+      for (Integer partitionId : partitionIds) {
+        ResultPartitionID resultPartitionID = resultPartitionMap.remove(partitionId);
+        if (resultPartitionID != null) {
+          resultPartitionIDS.add(resultPartitionID);
+        }
+      }
+
+      notifyStopTrackingPartitions(resultPartitionIDS, new AtomicInteger(MAX_RETRY_TIMES));
+    }
+
+    private void notifyStopTrackingPartitions(
+        Set<ResultPartitionID> partitionIDS, AtomicInteger remainingRetries) {
```

Comment:
Seems it's unnecessary to use AtomicInteger because another try will only be triggered after the previous fails.